src/vidpf.rs

```diff
@@ -0,0 +1,644 @@
+        let mut out = Self::zero(&Default::default());
+        zip(&mut out.0, zip(&lhs.0, &rhs.0))
+            .for_each(|(c, (a, b))| *c = <F as IdpfValue>::conditional_select(a, b, choice));
+        out
+    }
+}
+
+impl<F: FieldElement, const N: usize> ConditionallyNegatable for Weight<F, N> {
+    fn conditional_negate(&mut self, choice: Choice) {
+        self.0.iter_mut().for_each(|a| a.conditional_negate(choice));
+    }
+}
+
+impl<F: FieldElement, const N: usize> Add for Weight<F, N> {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let mut out = Self::zero(&Default::default());
+        zip(&mut out.0, zip(self.0, rhs.0)).for_each(|(c, (a, b))| *c = a.add(b));
+        out
+    }
+}
+
+impl<F: FieldElement, const N: usize> AddAssign for Weight<F, N> {
+    fn add_assign(&mut self, rhs: Self) {
+        zip(&mut self.0, rhs.0).for_each(|(a, b)| a.add_assign(b));
+    }
+}
+
+impl<F: FieldElement, const N: usize> Sub for Weight<F, N> {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        let mut out = Self::zero(&Default::default());
+        zip(&mut out.0, zip(self.0, rhs.0)).for_each(|(c, (a, b))| *c = a.sub(b));
+        out
+    }
+}
+
+impl<F: FieldElement, const N: usize> Encode for Weight<F, N> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        for a in self.0.iter() {
+            F::encode(a, bytes)?;
+        }
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(N * F::ENCODED_SIZE)
+    }
+}
+
+impl<F: FieldElement, const N: usize> Decode for Weight<F, N> {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let mut out = Self::zero(&Default::default());
+        for a in out.0.iter_mut() {
+            *a = F::decode(bytes)?;
+        }
+        Ok(out)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::io::Cursor;
+
+    use crate::{
+        codec::{Decode, Encode},
+        field::Field128,
+        idpf::IdpfValue,
+        vidpf::{EvalState, Key, PublicShare, Vidpf, VidpfInput, VidpfWeight, Weight},
+    };
+
+    type TestWeight = Weight<Field128, 3>;
+
+    fn vidpf_gen_setup(
+        input: &VidpfInput,
+        weight: &TestWeight,
+    ) -> (Vidpf<TestWeight>, [Key; 2], PublicShare<TestWeight>) {
+        let nonce = Vec::from("Mock Protocol uses a VIDPF");
+        let vidpf = Vidpf::<TestWeight>::new(nonce, ());
+        let (k0, k1, public) = vidpf.gen(input, weight).unwrap();
+        (vidpf, [k0, k1], public)
+    }
+
+    #[test]
+    fn vidpf_correctness_at_last_level() {
+        let input = VidpfInput::from_bytes(&[0xFF]);
+        let weight = TestWeight::new([21.into(), 22.into(), 23.into()]);
+        let (vidpf, [alice_key, bob_key], public) = vidpf_gen_setup(&input, &weight);
+
+        let alice_share = vidpf.eval(&alice_key, &public, &input).unwrap();
+        let bob_share = vidpf.eval(&bob_key, &public, &input).unwrap();
```

Comment:
For consistency with other test code in this repo.

Here and below.

```suggestion
        let value_share_0 = vidpf.eval(&alice_key, &public, &input).unwrap();
        let value_share_1 = vidpf.eval(&bob_key, &public, &input).unwrap();
```